tests/test_unit/test_prerun_transformations/test_transformations.py

```diff
@@ -0,0 +1,633 @@
+    ),
+    AddWaterInTransitTestCase(
+        case_id="AddWaterInTransitTestCase: shape key error 2", model={"foo": {"gate1": {"foo": 0}}}, error=KeyError
+    ),
+    AddWaterInTransitTestCase(
+        case_id="AddWaterInTransitTestCase: shape value error", model={"gate": {"gate1": {"foo": 0}}}, error=ValueError
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "test_case",
+    [pytest.param(test_case, id=test_case.case_id) for test_case in ADD_WATER_TEST_CASES],
+)
+def test_add_water_in_transit(cognite_client_mock: CogniteClient, test_case: AddWaterInTransitTestCase):
+
+    with pytest.raises(test_case.error) if test_case.error else nullcontext():
+        transformation = AddWaterInTransit(
+            discharge_ts_external_id=test_case.discharge_ts_external_id,
+            transit_object_type=test_case.transit_object_type,
+            transit_object_name=test_case.transit_object_name,
+        )
+
+        if test_case.discharge_start_time:
+            discharge_start_time = test_case.discharge_start_time
+        else:
+            discharge_start_time = test_case.shop_start_time - timedelta(hours=len(test_case.discharge_values))
+        discharge_times = pd.date_range(
+            start=discharge_start_time, periods=len(test_case.discharge_values), freq=test_case.time_frequency
+        )
+        discharge_times_ms = [time.replace(tzinfo=timezone.utc).timestamp() * 1000 for time in discharge_times]
+
+        inflow_times = pd.date_range(
+            start=test_case.shop_start_time, periods=len(test_case.inflow_values), freq=test_case.time_frequency
+        )
+        inflow_input_data = (pd.Series(test_case.inflow_values, index=inflow_times),)
+
+        start_time_ms = int(test_case.shop_start_time.replace(tzinfo=timezone.utc).timestamp() * 1000)
+        end_time = test_case.end_time if test_case.end_time else inflow_times[-1]
+        end_time_ms = int(end_time.replace(tzinfo=timezone.utc).timestamp() * 1000)
+
+        if test_case.model:
+            model = test_case.model
+        else:
+            model = {
+                test_case.transit_object_type: {
+                    test_case.transit_object_name: {test_case.shape_type: test_case.discharge_shape}
+                }
+            }
+
+        cognite_client_mock.time_series.data.retrieve.return_value = Datapoints(
+            external_id=test_case.discharge_ts_external_id,
+            value=test_case.discharge_values,
+            timestamp=discharge_times_ms,
+        )
+        cognite_client_mock.time_series.data.retrieve_latest.return_value = Datapoints(
+            external_id=test_case.discharge_ts_external_id,
+            value=[test_case.discharge_values[-1]],
+            timestamp=[discharge_times_ms[-1]],
+        )
+        cognite_client_mock.time_series.retrieve_multiple.return_value = [
+            TimeSeries(external_id=test_case.discharge_ts_external_id, is_step=True)
+        ]
```

Comment:
Not related to this PR, but: Would it make sense to put the discharge time series (datapoints) inside the time series tuple used in the apply method, to avoid these extra API calls for retrieving datapoints? Maybe not - it is only in the context of the transformation that we know which period it makes sense to fetch data for?